src/hugr.rs

```diff
@@ -73,6 +77,127 @@ impl Hugr {
         unimplemented!()
     }
 
+    /// Apply a simple replacement operation to the HUGR.
+    pub fn apply_simple_replacement(
+        &mut self,
+        r: SimpleReplacement,
+    ) -> Result<(), SimpleReplacementError> {
+        // 1. Check the parent node exists.
+        if self.nodes().all(|node| node != r.p) || self.get_optype(r.p).tag() != OpTag::Dfg {
+            return Err(SimpleReplacementError::InvalidParentNode());
+        }
+        // 2. Check that all the to-be-removed nodes are children of it and are leaves.
+        for node in &r.s {
+            if self.hierarchy.is_root(node.index)
+                || self.hierarchy.parent(node.index).unwrap() != r.p.index
+                || self.hierarchy.has_children(node.index)
+            {
+                return Err(SimpleReplacementError::InvalidRemovedNode());
+            }
+        }
+        // 3. Do the replacement.
+        // First locate the DFG in r.n. TODO this won't be necessary when we have DFG-rooted HUGRs.
+        let n_dfg_node =
+            r.n.nodes()
+                .find(|node: &Node| r.n.get_optype(*node).tag() == OpTag::Dfg)
+                .unwrap();
+        // 3.1. Add copies of all children of n_dfg_node to self. Exclude Input/Output nodes.
+        // Create map from old NodeIndex (in r.n) to new NodeIndex (in self).
+        let mut index_map: HashMap<NodeIndex, NodeIndex> = HashMap::new();
+        let n_indices =
+            r.n.hierarchy
+                .children(n_dfg_node.index)
+                .collect::<Vec<NodeIndex>>();
+        let n_sz = n_indices.len(); // number of replacement nodes including Input and Output
+        let n_non_io_indices = &n_indices[1..n_sz - 1]; // omit Input and Output
+        let self_input_node_index: NodeIndex = self.hierarchy.first(r.p.index).unwrap();
+        let n_output_node_index: NodeIndex = n_indices[n_sz - 1];
+        for n_index in n_non_io_indices {
+            let n_node = Node { index: *n_index };
+            let op: &OpType = r.n.get_optype(n_node);
+            let sig = op.signature();
+            if !sig.const_input.is_empty() {
+                return Err(SimpleReplacementError::InvalidReplacementNode());
+            }
+            let new_node_index = self.graph.add_node(sig.input.len(), sig.output.len());
+            self.op_types[new_node_index] = op.clone();
+            // Make r.p the parent
+            self.hierarchy
+                .insert_after(new_node_index, self_input_node_index)
+                .ok();
+            index_map.insert(*n_index, new_node_index);
+        }
+        // 3.2. Add edges between all newly added nodes matching those in n_dfg_node.
```

Comment:
IIUC the spec was written assuming implicit copies so I'm guessing this may change when that's ready.